File: ruby/373.find-k-pairs-with-smallest-sums.rb
def k_smallest_pairs(nums1, nums2, k)
  max_j = nums2.length

  nums1.each_with_object([]) do |n, heap|
    return heap if max_j.zero? && heap.length == k

    nums2.each_with_index do |m, j|
      break heap if j > max_j

      if (l = heap.bsearch_index { |pair| pair[0] + pair[1] > n + m })
        heap.insert(l, [n, m])
        heap.pop if heap.length > k
      elsif heap.length < k
        heap.push([n, m])
      else
        max_j = j
        break heap
      end
    end
  end
end

File: rust/src/373.find-k-pairs-with-smallest-sums.rs
use std::{cmp::Reverse, collections::BinaryHeap};

use crate::Solution;

#[allow(dead_code)]
impl Solution {
    pub fn k_smallest_pairs(nums1: Vec<i32>, nums2: Vec<i32>, k: i32) -> Vec<Vec<i32>> {
        let mut heap = BinaryHeap::new();
        let mut k = k as usize;
        let (m, n) = (nums1.len(), nums2.len());
        let mut ans = vec![];

        for i in 0..m {
            if heap.len() >= k {
                break;
            }

            heap.push(Reverse((nums1[i] + nums2[0], i, 0 as usize)));
        }

        while k > 0 {
            if let Some(Reverse((_sum, i, j))) = heap.pop() {
                ans.push(vec![nums1[i], nums2[j]]);

                if j + 1 < n {
                    heap.push(Reverse((nums1[i] + nums2[j + 1], i, j + 1)));
                }
            }
            k -= 1;
        }
        ans
    }
}
